tags/TagMapContext.md

```diff
@@ -0,0 +1,144 @@
+# Summary
+A `Tag` is a key-value pair that can be used to label anything that is associated
+with a specific operation, such as an HTTP request. These `Tag`s are used to
+aggregate measurements in a [`View`](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/DataAggregation.md#view) 
+according to unique value of the `Tag`s. The `Tag`s can also be used to filter (include/exclude)
+measurements in a `View`.
+
+# TagMap 
+`TagMap` is an abstract data type that represents collection of tags. 
+i.e., each key is associated with exactly one value, but multiple keys can be
+associated with the same value.  `TagMap` is serializable, and it represents
+all of the information that could be propagated across process boundaries.
+
+## TagKey
+
+A string or string wrapper, that contains two parts, scope and key. The scope
+of a tag is used to determine if a tag should be propagated or not.
+
+
+**TagKey Syntax:**  
+Scope and key are separated by ':'.
+```
+<Scope>:<Key>
+```
+
+**Scope**
+Scope of a `TagKey` defines a boundary within which the `TagKey` is relevant. For example 
+a `TagKey` http_route on frontend service could have a scope that spans across multiple services
+that contributes to fulfilling the http request. On the other hand if you are measuring latency
+of key individual functions in a ProductCatalog service then the scope of a `TagKey` func_name
+is most likely limited to a process in ProductCatalog service.
+
+The scope is used to determine whether a Tag should be propagated or not.
+
+**Key**   
+Key is a string, with following restrictions:
+
+- Must contain only printable ASCII (codes between 32 and 126 inclusive with below Exception)
+- - Exceptions ':', ','
+- Must have length greater than zero and less than 256 combined with the scope.
+- Must not be empty.
+
+### Scope
+
+#### Scope Syntax
+
+```
+<scope-type1-code>,<scope-value1>,<scope-type2-code>,<scope-value2>,...
+```
+- Scope MUST always be in a Type-Value pair.
+
+There are following scopes types.
+
+#### Scope Types
+Scope Type is represented using code to reduce the size.
+
+| Code | Type  |
+|------|-------|
+| 0    | Process |
+| 1    | Internal-Service|
+| 2    | External-Service|
+
+
+##### Process
+A `TagKey` with the scope type *Process* is relevant only to stats collected by the process.
+Hence it should not be propagated across different processes. This is the default scope. If no
+scope is present in the `Tagkey` then it is treated as of Process scope type.
+
+##### Internal-Service
```

Comment:
You kind of make service a first class concept in OC by adding this. I see the value but I think we should allow users to achieve this in a different way (control the incoming/outgoing tags in the http/grpc plugins).

Reply:
I think the library should be aware of the Scope to do the correct thing in terms of propagating or not these tags.

Reply:
This one is very confusing to me. First, the description is not particularly clear about what "internal service" means. But more importantly, how can the instrumentation library know about that? E.g. say I am making a request to some host:port, how can OC tell that it's a request to "internal service"?

Reply:
I have re-drafted the spec based on suggestion and removed the concept of internal/external service. I will post the new draft soon.